crates/bevy_ecs/src/schedule/executor/single_threaded.rs

```diff
@@ -85,6 +86,21 @@ impl SystemExecutor for SingleThreadedExecutor {
 
             should_run &= system_conditions_met;
 
+            let system = &mut schedule.systems[system_index];
+
+            // SAFETY:
+            // - The caller ensures that `world` has permission to read any data
+            //   required by the system.
+            // - `update_archetype_component_access` has been called for system.
```

Comment:
This is the same requirement for `get_param` which happens inside every `System::run_unsafe`,
since `validate_param` is essentially `get_param` with read-only restrictions, it inherits the same safety.
I have no clue why it's never called for single threaded executors, I raised this in discord but didn't get attention

Reply:
Oh, I remember this: `update_archetypes_component_access()` gets called by `System::run()`.  The single-threaded executors have `&mut World`, so they can call `run()` instead of `run_unsafe()` and have this handled automatically.  

Hmm, should `System::run()` check `validate_params_unsafe()`?  That would make this change work with `World::run_system()` and friends, and it would remove the need for changes to the simple executor.  

Another option is a safe `validate_params()` that takes `&mut World` and calls `update_archetype_component_access()` itself.

Reply:
> I have no clue why `update_archetype_component_access` is never called for single threaded executors, I raised this in Discord but didn't get attention.

Were you referring to `main` here?
On `main`, the simple/single-threaded executors call the safe `run` function for all systems, including conditions.

https://github.com/bevyengine/bevy/blob/e3b6b125a0a3b9bd15d939b51136cc186f086c56/crates/bevy_ecs/src/system/system.rs#L77-L86

Reply:
okay, this complicates `validate_param`

Reply:
Moving this to discord discussion
https://discord.com/channels/691052431525675048/1287411721593426031

Reply:
Agreed on duplicate `update_archetypes_component_access` for safe variant of `validate_param`